File: Algorithms/p095_Unique_Binary_Search_Trees_II/p095_Unique_Binary_Search_Trees_II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        if n==0:
            return []
        table = [[] for i in range(n+1)]
        table[0].append(None)
        for i in range(1,n+1):
            for j in range(i):
                for t1 in table[j]:
                    for t2 in table[i-1-j]:
                        root = TreeNode(0)
                        idx = [0]
                        root.left = self.copyTree(t1,idx)
                        idx[0] += 1
                        root.val = idx[0]
                        root.right = self.copyTree(t2,idx)
                        table[i].append(root)
        return table[n]

    def copyTree(self, root, idx):
        if not root:
            return None
        root1 = TreeNode(0)
        root1.left = self.copyTree(root.left,idx)
        idx[0] += 1
        root1.val = idx[0]
        root1.right = self.copyTree(root.right,idx)
        return root1

File: Algorithms/p095_Unique_Binary_Search_Trees_II/p095_Unique_Binary_Search_Trees_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if(n==0) return {};
        vector<vector<TreeNode*>> table(n+1);
        table[0].push_back(0);
        for(int i=1;i<=n;++i){
            for(int j=0;j<i;++j){
                for(auto t1 : table[j]){
                    for(auto t2 : table[i-1-j]){
                        TreeNode* root = new TreeNode(0);
                        int id = 0;
                        root->left = copyTree(t1,id);
                        root->val = ++id;
                        root->right = copyTree(t2,id);
                        table[i].push_back(root);
                    }
                }
            }
        }
        return table[n];
    }
    TreeNode* copyTree(TreeNode* root, int& id) {
        if(!root) return 0;
        TreeNode* root1 = new TreeNode(0);
        root1->left = copyTree(root->left,id);
        root1->val = ++id;
        root1->right = copyTree(root->right,id);
        return root1;
    }
};
